### DISCUSSION TITLE
prefetching

### DISCUSSION THREAD

**Discussion by nicolodavis** [urql-graphql/urql#2450] [Q&A]
I have a schema (simplified) like:

```graphql
type Project {
	id: ID!
	components: [Component!]!
}

type Component {
	id: ID!
	name: String!
}
```

I would like to avoid having the `components` resolver do fetches when used, so I prefetch the whole state when I fetch the `Project` by:

1. Storing it in a raw `json` field:

```graphql
type Project {
	id: ID!
	components: [Component!]!
        json: JSON!    // <-----
}
```

2. Using local resolvers to resolve the `components` link (this local resolver reads from the `json` field and uses the already cached `Component` entities as resolution targets).

I'm wondering if there is a better pattern to solve this problem of requiring one large fetch at the start of the application and having everything else after that served from the cache. In my specific solution I have this problem:

If my initial query is something like:

```
query {
  project {
    id
    json
  }
  components {  // <-- in order to cache the Component entities
    id
    name
  }
}
```

then a subsequent query like:

```
query {
  project {
    id
    json
    components {
      id
      name
    }
  }
}
```

actually triggers a fetch, even though it has all the information it needs to serve this from the cache (it has the `Component` entities cached and it also has the links cached in the `json` field).

**Comment by nicolodavis** [comment]
**EDIT**: Turns out that the spurious fetch was caused due to a bug in one of my cache resolvers.

My larger question about whether this is a good pattern for prefetching still stands.

**Reply by frederikhors** [reply]
But why the big json field? Can't you just query what do you need?

**Reply by nicolodavis** [reply]
The json field isn't particularly big. I use it so that I can perform the linking of entities on the client using local resolvers.

For example, the raw data (on the backend) for a project might look like:
```js
{
  id: "project1",
  components: ["c1", "c2"],
}

{
  id: "c1",
  name: "Component 1",
}

{
  id: "c2",
  name: "Component 2",
}
```

In order to link the list of component ID's to the actual component entities, I need to pass the list of ID's to the client. I do this via the `json` field.

```js
{
  query {
    project {
      id
      json
    }
  }
}

{
  id: "project1",
  json: {
    components: ["c1", "c2"],
  }
}
```

Then, on the client I can use a local resolver to resolve those two component ID's to the actual components.

The alternative would be to do all the resolving on the backend, but having the raw data on the client side is useful to me for other reasons specific to my app which is why I also use it to resolve links via local resolvers.

**Reply by nicolodavis** [reply]
My understanding is that a local resolver only has access to retrieved data that is compatible with the GraphQL schema (through the `parent` and `cache` objects). If there is a way to access raw data similar to what a backend resolver would see, I'd love to learn how to avoid this `json` dance altogether.
